esp/esp/program/controllers/autoscheduler/controller.py

```diff
@@ -0,0 +1,394 @@
+            row = []
+            if action["action"] == "schedule":
+                row.append([u"Scheduled {}:".format(
+                                self.section_identifier(action["section"])),
+                           self.section_info(action["section"])])
+                row.append([u"In {}:".format(
+                                self.roomslot_identifier(
+                                    action["start_roomslot"])),
+                           self.roomslot_info(action["start_roomslot"])])
+            elif action["action"] == "move":
+                row.append([u"Moved {}:".format(
+                                self.section_identifier(action["section"])),
+                           self.section_info(action["section"])])
+                row.append([u"From {}:".format(
+                                self.roomslot_identifier(
+                                    action["prev_start_roomslot"])),
+                           self.roomslot_info(action["prev_start_roomslot"])])
+                row.append([u"To {}:".format(
+                                self.roomslot_identifier(
+                                    action["start_roomslot"])),
+                           self.roomslot_info(action["start_roomslot"])])
+            elif action["action"] == "unschedule":
+                row.append([u"Unscheduled {}:".format(
+                                self.section_identifier(action["section"])),
+                           self.section_info(action["section"])])
+                row.append([u"From {}:".format(
+                                self.roomslot_identifier(
+                                    action["prev_start_roomslot"])),
+                           self.roomslot_info(action["prev_start_roomslot"])])
+            elif action["action"] == "swap":
+                sec1, sec2 = action["sections"]
+                rs2, rs1 = action["original_roomslots"]
+                row.append([u"Swapped {}:".format(
+                                self.section_identifier(sec1)),
+                           self.section_info(sec1)])
+                row.append([u"Now in {}:".format(
+                                self.roomslot_identifier(rs1)),
+                           self.roomslot_info(rs1)])
+                row.append([u"and {}:".format(
+                                self.section_identifier(sec2)),
+                           self.section_info(sec2)])
+                row.append([u"Now in {}:".format(
+                                self.roomslot_identifier(rs2)),
+                           self.roomslot_info(rs2)])
+            else:
+                raise SchedulingError("History was broken.")
+            rows.append(row)
+        if len(rows) > 0:
+            final_scores, final_total_score = \
+                self.optimizer.manipulator.scorer.get_all_score_schedule()
+            diffs = []
+            for (scorer, weight, old_score), (scorer2, weight, new_score) in \
```

Comment:
you assign to "weight" twice? maybe name them weight and weight2 and assert that they're equal?